pkg/sql/sql_activity_update_job.go

```diff
@@ -0,0 +1,688 @@
+// rows selected to be inserted into the activity tables
+var sqlStatsActivityTopCount = settings.RegisterIntSetting(
+	settings.TenantWritable,
+	"sql.stats.activity.top.max",
+	"the limit per column for the top number of statistics to be flushed "+
+		"to the activity tables",
+	500,
+	settings.NonNegativeInt,
+)
+
+// sqlStatsActivityMaxPersistedRows specifies maximum number of rows that will be
+// retained in system.statement_statistics and system.transaction_statistics.
+// Defaults computed 500(top limit)*6(num columns)*24(hrs)*3(days)=216000
+// to give a minimum of 3 days of history. It was rounded down to 200k to
+// give an even number. The top 500(controlled by sql.stats.activity.top.max)
+// are likely the same for several columns, so it should still give 3 days
+// of history for the default settings
+var sqlStatsActivityMaxPersistedRows = settings.RegisterIntSetting(
+	settings.TenantWritable,
+	"sql.stats.activity.persisted_rows.max",
+	"maximum number of rows of statement and transaction"+
+		" activity that will be persisted in the system tables",
+	200000, /* defaultValue*/
+	settings.NonNegativeInt,
+).WithPublic()
+
+const numberOfTopColumns = 6
+
+type sqlActivityUpdateJob struct {
+	job *jobs.Job
+}
+
+// Resume implements the jobs.sqlActivityUpdateJob interface.
+// The SQL activity job runs AS a forever-running background job
+// and runs the SqlActivityUpdater according to sql.stats.activity.flush.interval.
+func (j *sqlActivityUpdateJob) Resume(ctx context.Context, execCtxI interface{}) (jobErr error) {
+	log.Infof(ctx, "starting sql stats activity flush job")
+	// The sql activity update job is a forever running background job.
+	// It's always safe to wind the SQL pod down whenever it's
+	// running, something we indicate through the job's idle
+	// status.
+	j.job.MarkIdle(true)
+
+	execCtx := execCtxI.(JobExecContext)
+	stopper := execCtx.ExecCfg().DistSQLSrv.Stopper
+	settings := execCtx.ExecCfg().Settings
+	statsFlush := execCtx.ExecCfg().InternalDB.server.sqlStats
+	metrics := execCtx.ExecCfg().JobRegistry.MetricsStruct().JobSpecificMetrics[jobspb.TypeAutoUpdateSQLActivity].(activityUpdaterMetrics)
+
+	flushDoneSignal := make(chan struct{})
+	defer func() {
+		statsFlush.SetFlushDoneCallback(nil)
+		close(flushDoneSignal)
+	}()
+
+	for {
+		statsFlush.SetFlushDoneCallback(func() {
+			flushDoneSignal <- struct{}{}
+		})
+		select {
+		case <-flushDoneSignal:
+			// A flush was done. Set the timer and wait for it to complete.
+			if enabled.Get(&settings.SV) {
+				updater := NewSqlActivityUpdater(settings, execCtx.ExecCfg().InternalDB)
+				if err := updater.TransferStatsToActivity(ctx); err != nil {
+					log.Warningf(ctx, "error running sql activity updater job: %v", err)
+					metrics.numErrors.Inc(1)
+				}
+			}
+		case <-ctx.Done():
+			return nil
+		case <-stopper.ShouldQuiesce():
+			return nil
+		}
+	}
+}
+
+type activityUpdaterMetrics struct {
+	numErrors *metric.Counter
```

Comment:
you must export numErrors (capitalize) if you want those to show up as metrics.  MetricStruct ignore any unexported metrics.